diff --git a/src/kosmickrisp/vulkan/kk_physical_device.c b/src/kosmickrisp/vulkan/kk_physical_device.c
index 5b981faaa69..5f44ac4ba1b 100644
--- a/src/kosmickrisp/vulkan/kk_physical_device.c
+++ b/src/kosmickrisp/vulkan/kk_physical_device.c
@@ -210,7 +211,9 @@ kk_get_device_features(
       .drawIndirectFirstInstance = true,
       .dualSrcBlend = true,
       .fragmentStoresAndAtomics = true,
+      .fillModeNonSolid = true, // HACK for DXVK
       .fullDrawIndexUint32 = true,
+      .geometryShader = true, // HACK for DXVK
       .imageCubeArray = true,
       .independentBlend = true,
       .inheritedQueries = true,
